bittide/src/Bittide/Axi4.hs

```diff
@@ -28,89 +46,708 @@ import Bittide.Extra.Maybe
+  forall dom addedWidth idWidth destWidth userType .
   ( HiddenClockResetEnable dom
-  , KnownNat dataWidth
-  , 1 <= dataWidth
+  , KnownNat addedWidth
+  , KnownNat idWidth
+  , KnownNat destWidth
   , Eq userType
   , NFDataX userType) =>
   Circuit
     (Axi4Stream dom ('Axi4StreamConfig 1 idWidth destWidth) userType)
-    (Axi4Stream dom ('Axi4StreamConfig dataWidth idWidth destWidth) userType)
-axisFromByteStream = Circuit (mealyB go initState)
+    (PackedAxi4Stream dom
+      ('Axi4StreamConfig (addedWidth + 1) idWidth destWidth)
+      (Vec (addedWidth + 1) userType))
+axisFromByteStream = Circuit (mealyB go Nothing)
  where
-  initState :: (Index dataWidth, Vec (dataWidth - 1) (Unsigned 8, Bool, Bool))
-  initState = (0, repeat initTempAxi)
-  initTempAxi = (0, False, False)
-  go state (Nothing, _) = (state, (Axi4StreamS2M{_tready = True}, Nothing))
-  go (counter, storedBytes) (Just Axi4StreamM2S{..}, Axi4StreamS2M{_tready}) =
-    ((newCounter, nextStoredBytes), (smallS2M, bigM2S))
+  go axiStored (axiIn, Axi4StreamS2M{_tready}) = (axiNext, (lhsS2M, rhsM2S))
    where
-    smallS2M = Axi4StreamS2M{_tready = smallReady}
-    bigM2S
-      | not upscaleDone = Nothing
-      | otherwise       = Just $ Axi4StreamM2S
-        { _tdata = newData
-        , _tkeep = newKeeps
-        , _tstrb = newStrobes
-        , _tuser = _tuser
-        , _tid   = _tid
-        , _tdest = _tdest
-        , _tlast = _tlast}
-    (newData, newKeeps, newStrobes) = unzip3 intermediateBytes
-
-    upscaleDone = counter == maxBound || _tlast
-    smallReady  = not upscaleDone || _tready
-    newCounter
-      | upscaleDone && _tready && _tready = 0
-      | upscaleDone = counter
-      | otherwise   = satSucc SatWrap counter
-
-    -- TODO: Replace this relatively expensive replace operation with a shift operation
-    intermediateBytes = replace counter (head _tdata, head _tkeep, head _tstrb)
-      (storedBytes :< initTempAxi)
-
-    nextStoredBytes
-      | upscaleDone && _tready = repeat initTempAxi
-      | otherwise              = init intermediateBytes
-
-{-# NOINLINE axisToByteStream #-}
+    undefUser = deepErrorX "axisFromByteStream: _tuser undefined"
+    -- Try to append the incoming axi to the stored axi
+    combinedAxi = axiUserMap (\(v1, v2) -> v1 :< v2) <$> concatAxi axiStored axiIn
+    inpValid = isJust axiIn && isJust combinedAxi
+    inpInvalid = isJust axiIn && isNothing combinedAxi
+
+    -- Shifting bytes
+    axiPreShift = combinedAxi <|> (axiUserMap extendWithErrorX . extendAxi <$> axiStored)
+    axiBytes =
+      zipWith4 (\d k s u -> orNothing k (d, s, u))
+      (maybe (repeat 0) _tdata axiPreShift)
+      (maybe (repeat False) _tkeep axiPreShift)
+      (maybe (repeat False) _tstrb axiPreShift)
+      (maybe (repeat undefUser) _tuser axiPreShift)
+
+    (newData, newKeeps, newStrobes, newUsers) = unzip4 $
+      (maybe (0, False, False, undefUser) (\(d, s, u) -> (d, True, s, u))) <$> shiftPackVec axiBytes
```

Comment:
This `shiftPackVec` seems expensive, when we know that only the `axiIn`  should contain a `tkeep` that's deasserted.

I guess it's acceptable for now. But do create an issue on the  issue tracker that this function should be made more resource efficient.